Fixed #27391 -- Implemented SimpleTestCase.debug().

asfaltboy opened the pull request:
https://code.djangoproject.com/ticket/27391

Users calling debug() expect it to bubble up an exception if occurring in test, but behave the
same as run() when no exception occurred.

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
Did you consider trying to factor the shared logic with `__call__()` to avoid the repetition?

asfaltboy commented:
I was really trying to avoid it, I'd hate to see current simplicity get bogged down... but, how about this:

```python
def setup_and_call(self, result=None, debug=False):
    """Wrapper around default __call__ method to perform common Django test
    set up. This means that user-defined Test Cases aren't required to
    include a call to super().setUp().

    If passed debug=True, does does not catch errors and calls super().debug
    instead of __call__.
    """
    @contextmanager
    def catch_or_bubble(debug):
        if debug:
            yield
        try:
            yield
        except Exception:
            result.addError(self, sys.exc_info())
            return

    testMethod = getattr(self, self._testMethodName)
    skipped = (
        getattr(self.__class__, "__unittest_skip__", False) or
        getattr(testMethod, "__unittest_skip__", False))

    if not skipped:
        with catch_or_bubble(debug):
            self._pre_setup()
    if debug:
        super(SimpleTestCase, self).debug()
    else:
        super(SimpleTestCase, self).__call__(result)

    if not skipped:
        with catch_or_bubble(debug):
            self._post_teardown()

```

asfaltboy commented:
@timgraham please take a look at [the updated code][1]. Is this more or less what you envisioned? Or would you like anything changed here? I think I will remove the first paragraph the helper method's docstring, it's repetitive.

[1]: https://github.com/asfaltboy/django/blob/d1f2eae8a86e3f7d2bbcf55661fd1590ae955f0e/django/test/testcases.py#L193-L242

timgraham commented on tests/test_utils/test_simpletestcase.py:
```diff
@@ -0,0 +1,84 @@
+        test_suite.addTest(case)
+        test_suite.run(self.get_runner()._makeResult())
+        # _post_teardown removes the object added in _pre_setup()
+        self.assertEqual(case._running_test, [])
```
Did you consider using mocking and assertions such as `assert_called_once_with()`? This might be a bit cleaner and more self-explanatory than introspecting `_running_test`.

timgraham commented:
@asfaltboy, do you have any time and interest in completing this soon?

asfaltboy commented:
Hi @timgraham sorry for taking long to respond (and for hogging the CI machine with amends).

How does the test case look now?

timgraham commented:
It looks like there's a test isolation problem (see the build failures).

asfaltboy commented:
@timgraham it took "a bit" longer than promised, but I finally got to it, the isolation issue was due to raised exception in Debug requiring to properly cleaning up after we're done with suite.

It's not a coincidence, because this is exactly what the fix intends to do, to let the test running framework (e.g pytest) to be able to catch errors and do their own cleanup as desired.

timmartin commented on django/test/testcases.py:
```diff
@@ -202,13 +216,20 @@ def __call__(self, result=None):
                 self._pre_setup()
             except Exception:
                 result.addError(self, sys.exc_info())
```
This will fail if `result` is `None`.

asfaltboy replied:
Good catch! Thanks @timmartin I'll review this later (probably will add another test for this case and make sure we avoid calling `result.addError` when it's `None`)

asfaltboy replied:
Hmm, it seems the default `result=None` kwarg and the call to `result.addError` are not added in this PR.

So, strictly speaking, this as an unrelated issue and, we'd probably be better discussing this in a new issue and/or patch.

However, I went ahead and mimicked [the way `unittest.suite` defines `TestSuite.debug()`][1], by making the result non-optional in our `_setup_and_call` method and by using `_DebugResult()` in our `debug()` method.

Existing method `__call__` is untouched, allowing kwarg of `result=None`, since I'm not sure whether this is used by anyone. If there are any users of this code, they will also stumble on the `AttributeError` that will be raised when calling `result.addError` above. BTW: the test suite passes by making `result` a required parameter to `__call__`, so if this is a desired change let me know and I'll update the `__call__` kwarg as well.

[1]: https://github.com/python/cpython/blob/c4d9df5fd719ad08e68e0950ce22a80f43e4f35d/Lib/unittest/suite.py#L135-L138

asfaltboy commented:
About updating release notes, I think we can rebase this PR over latest master and add this under _2.0.txt: Miscellaneous_ (currently line 634):

```patch
+* Implement the ``SimpleTestCase.debug()`` method to allow a test to run without
+  collecting the result. This allows exceptions raised by the test to be propagated to
+  the caller, and is used to support running tests under a debugger.
+  (:ticket:`27391`).
```

timgraham commented on django/test/testcases.py:
```diff
@@ -191,6 +192,21 @@ def __call__(self, result=None):
         set up. This means that user-defined Test Cases aren't required to
         include a call to super().setUp().
         """
+        self._setup_and_call(result, debug=False)
```
Remove `debug=False` since that's the default.

asfaltboy replied:
addressed in 3bf9e42

timgraham commented on django/test/testcases.py:
```diff
@@ -191,6 +192,21 @@ def __call__(self, result=None):
+        Perform the following in order: pre-setup, run test, post-teardown,
+        skipping pre/post hooks if test is set to be skipped.
+
+        If passed debug=True, reraise any errors in setup and use
```
removed "passed"

asfaltboy replied:
addressed in 3bf9e42

timgraham commented on tests/test_utils/test_simpletestcase.py:
```diff
@@ -0,0 +1,90 @@
+        _post_teardown.assert_called_once_with()
+
+    def test_debug_bubbles_exception(self, _pre_setup, _post_teardown):
+        """debug() should bubble up exceptions before any cleanup."""
```
remove "should", e.g. "debug() bubbles up exceptions before cleanup."

asfaltboy replied:
addressed in 3bf9e42

timgraham commented on tests/test_utils/test_simpletestcase.py:
```diff
@@ -0,0 +1,90 @@
+        with self.assertRaisesMessage(Exception, 'Should bubble up when using debug()'):
+            # below is same as test_suite.debug()
+            result = _DebugResult()
+            test_suite.run(result, True)
```
Pass `True` as a keyword argument for better readability.

asfaltboy replied:
addressed in 3bf9e42

timgraham commented on tests/test_utils/test_simpletestcase.py:
```diff
@@ -0,0 +1,90 @@
+        test_suite._handleModuleTearDown(result)
+
+    def test_debug_cleanup(self, _pre_setup, _post_teardown):
+        """ With no errors, cleanup should occur as usual. """
```
Remove extra spaces around docstring.

asfaltboy replied:
addressed in 3bf9e42

asfaltboy commented:
Hi @timmartin @timgraham , I've had some time during the europython event to address the requested changes. I've had to move the addError call to under the raise call in both pre-set and post-teardown clauses (which makes sense) and added tests for those.

The sqlite3-windows tests were failing on CI so I rebased (maybe un-necessarily) and now [the sqlite3-bionic test is failing](https://djangoci.com/job/pull-requests-bionic/database=sqlite3,label=bionic-pr,python=python3.6/680/console). Is this unrelated?

timgraham commented:
Now I can comment both `result.addError(self, sys.exc_info())` without seeing any test failures. 

Yes, the build failure is unrelated. Don't worry about that.

Ian-Foote commented:
This probably needs rebasing first, so it can be re-tested.

asfaltboy commented:
@Ian-Foote @LuRsT thanks for your interest in merging this PR, it motivated me to get back to it.

I went ahead and rebased over latest master, and addressed the missing tests mentioned by @timgraham. These tests validate that the existing TestCase behavior, of capturing and logging errors in `pre_setup` & `post_teardown`, is retained after this change. (waiting for CI to confirm this runs correctly as tests fail on my machine fail erratically)

Is there anything we can address? Shall I squash the older WIP commits or leave them there? If nothing else, shall we mention this feature in release notes or docs?

asfaltboy commented:
The failing test seems to be unrelated as it failed to check out the "PR merge commit", maybe this is an issue with the git plugin like [the one described here](https://issues.jenkins-ci.org/browse/JENKINS-26290?focusedCommentId=269569&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-269569) ?

Ian-Foote commented:
@carltongibson @felixxm

felixxm commented:
@asfaltboy Failing tests are related with force-push you don't need to worry about them.

It is a new feature so documentation (`topics/testing/tools.txt`) and release notes are required.

asfaltboy commented:
Thanks @felixxm, I doc'd the method in `topics/testing/tools.txt` and in `releases/3.0.txt`. 

I feel, however, that this may be "too low-level" for the tools docs.
The target audience is Django test tool developers (pytest-django, nose, etc...), rather than Django users. I see this as the same reason that we don't document `SimpleTestCase.run()`, or am I wrong and we should have documented `run()` in the first place 🙂?

auvipy commented on docs/topics/testing/tools.txt:
```diff
@@ -775,6 +775,16 @@ If your tests make any database queries, use subclasses
+
+    This method can be used by test runners to implement their own exception handling, for example when running under a debugger. Under the hood this method calls :meth:`unittest.TestCase.debug` in the same place it would normally call :meth:`unittest.TestCase.run()`.
+
+.. versionadded:: 3.0
```
3.1

asfaltboy replied:
amended last commit (see a56357ccf0)

auvipy commented on docs/releases/3.0.txt:
```diff
@@ -365,6 +366,11 @@ Tests
 * Django test runner now supports a ``--pdb`` option to spawn a debugger at
   each error or failure.
 
+* Implement the :meth:`~django.test.SimpleTestCase.debug()` method to
```
this should be moved to 3.1

asfaltboy replied:
amended last commit (see a56357ccf0)

felixxm commented:
@asfaltboy I moved extra tests to a separate PR #11936.

felixxm commented:
@asfaltboy Thanks for updates :+1: I rebased from master, squashed commits. removed the `SimpleTestCase.debug()` documentation, added a `versionchanged` admonition, updated release notes and pushed minor edits. It's ready :rocket:
